core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -255,17 +273,162 @@ private void rebuildIndex() throws Exception {
       }
    }
 
+   // Initialise missing internal metadata state for corrupt data
+   private PrivateMetadata generateMissingInternalMetadata() {
+      // Optimistic Transactions
+      AdvancedCache<?, ?> cache = ctx.getCache().getAdvancedCache();
+      Configuration config = cache.getCacheConfiguration();
+      TransactionConfiguration txConfig = config.transaction();
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
```

Comment:
We should probably only initialize this builder if one of the ifs below are true. Otherwise all entries will have an empty internal metadata taking up space it doesn't need to.

Reply:
this is only invoked when `fe.internalMetadataLen > 0`. 
It is never empty :)

Reply:
And we are 100% sure you can only have internal metadata with optimistic or async xsite?

Reply:
I added the internal metadata for those 2 configurations. Not sure if somebody added it for something else :) 
in any case, the Builder is smart enough to detect if the internal metadata is null

Reply:
That isn't the problem though. I was more referring to we will marshall the EMPTY PrivateMetadata for every object where neither is true. But it sounds like there should be no other uses of internal metadata.